### ISSUE TITLE
SSH Dialer can hang forever.

### ISSUE THREAD

**Issue by cjcullen** [kubernetes/kubernetes#23835]
The ssh default timeout is "0" (no timeout). This appears to cause problem for ssh tunnels. A tunnel open attempt can hang indefinitely, and we don't try to kill in-progress open attempts. Most of the time, a failed open attempt will time out after the standard 127 second TCP timeout, but it appears that sometimes it hangs after the TCP connection is established, in the TLS handshake.

**Comment by a-robinson** [comment]
Looks like the SSH timeout was just added to crypt/ssh 10 days ago, so it isn't in any released version of go yet...
https://github.com/golang/crypto/commit/c7e3b0ebdd409a0d024e3d71801427ab0e05fb2e

**Comment by a-robinson** [comment]
And CJ's stack trace makes it clear that the problem isn't actually in net.Dial hanging, but ssh.NewClientConn, which relies on net.Dial having successfully returned.

I'll open a new issue against the crypt/ssh package, but it looks like we may have to hack our own timeout logic in on top of ssh.Dial.

```
goroutine 85851 [IO wait, 178 minutes]:
net.(*pollDesc).Wait(0xc20ae2e840, 0x72, 0x0, 0x0)
	/usr/src/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc20ae2e840, 0x0, 0x0)
	/usr/src/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc20ae2e7e0, 0xc20c38b2d0, 0x1, 0x1, 0x0, 0x7f20600e52a0, 0xc20c38b2d8)
	/usr/src/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc20b7add18, 0xc20c38b2d0, 0x1, 0x1, 0x0, 0x0, 0x0)
	/usr/src/go/src/net/net.go:121 +0xdc
io.ReadAtLeast(0x7f20600eb598, 0xc20b7add18, 0xc20c38b2d0, 0x1, 0x1, 0x1, 0x0, 0x0, 0x0)
	/usr/src/go/src/io/io.go:298 +0xf1
io.ReadFull(0x7f20600eb598, 0xc20b7add18, 0xc20c38b2d0, 0x1, 0x1, 0x40, 0x0, 0x0)
	/usr/src/go/src/io/io.go:316 +0x6d
golang.org/x/crypto/ssh.readVersion(0x7f20600eb598, 0xc20b7add18, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/Godeps/_workspace/src/golang.org/x/crypto/ssh/transport.go:303 +0x167
golang.org/x/crypto/ssh.exchangeVersions(0x7f205ff485c0, 0xc20b7add18, 0xc20c38b2c0, 0xa, 0x10, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/Godeps/_workspace/src/golang.org/x/crypto/ssh/transport.go:287 +0x2f1
golang.org/x/crypto/ssh.(*connection).clientHandshake(0xc20cdb9100, 0xc20bdf6380, 0x11, 0xc20b56e1e0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/Godeps/_workspace/src/golang.org/x/crypto/ssh/client.go:91 +0x132
golang.org/x/crypto/ssh.NewClientConn(0x7f20600ea500, 0xc20b7add18, 0xc20bdf6380, 0x11, 0xc20b56e140, 0x0, 0x0, 0x0, 0xe, 0x0, ...)
	/go/src/k8s.io/kubernetes/Godeps/_workspace/src/golang.org/x/crypto/ssh/client.go:74 +0x140
golang.org/x/crypto/ssh.Dial(0x1d48080, 0x3, 0xc20bdf6380, 0x11, 0xc20b56e140, 0x11, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/Godeps/_workspace/src/golang.org/x/crypto/ssh/client.go:176 +0xf9
k8s.io/kubernetes/pkg/ssh.(*SSHTunnel).Open(0xc20d0490e0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/ssh/ssh.go:114 +0xbc
k8s.io/kubernetes/pkg/ssh.(*SSHTunnelList).createAndAddTunnel(0xc208228360, 0xc2085c6720, 0xe)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/ssh/ssh.go:433 +0x331
created by k8s.io/kubernetes/pkg/ssh.(*SSHTunnelList).removeAndReAdd
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/ssh/ssh.go:345 +0x3ae
```

**Comment by a-robinson** [comment]
Given how hard this looks like it should be to trigger, it feels more like a P1 than a P0. Although we've seen it twice in the last couple days alone...

**Comment by a-robinson** [comment]
Opened https://github.com/golang/go/issues/15113

**Comment by a-robinson** [comment]
CJ will add in some sort of timing out in our code as a stop-gap until the underlying issue can be fixed in crypt/ssh. It'll involve leaking the stuck goroutines, but is better than apiservers getting into an unrecoverable state of having no working tunnels.

**Comment by cjcullen** [comment]
Fixed by #23843. Released in version 1.2.2.
